### PR TITLE
sql: disallow aggregate funcs in RETURNING

### LINKED ISSUE #6705: sum() in insert returning clause causes crash
Please follow the steps below to help us help you.

1. JESSEs-MacBook-Pro:cockroach jesseseldess$ grep -F '[config]' cockroach-data/logs/cockroach.INFO
I160516 13:34:10.154643 util/log/clog.go:998  [config] file created at: 2016/05/16 13:34:10
I160516 13:34:10.154643 util/log/clog.go:998  [config] running on machine: JESSEs-MBP
I160516 13:34:10.154643 util/log/clog.go:998  [config] binary: CockroachDB beta-20160512-89-gd66b5e9 (darwin amd64, built 2016/05/16 15:41:41, go1.6)
I160516 13:34:10.154643 util/log/clog.go:998  [config] arguments: [./cockroach start --background --background=false]

2. Please describe the issue you observed:

- What did you do?
In the `RETURNING` clause of an `INSERT`, I tried to use the `sum()` function to sum up the values of two inserted decimals. I later learned that this isn't possible, but instead of a crash, we should show an error message. Also, it's interesting to note that other functions used in this way don't cause crashes, but they also don't do anything (should they return errors?).

```
root@:26257> INSERT INTO accounts (id, balance) VALUES (1, 5000.50), (2, 7500.50) RETURNING sum(balance);
```

- What did you expect to see?
The sum of 5000.50 and 7500.50.

- What did you see instead?

```
root@:26257> INSERT INTO accounts (id, balance) VALUES (1, 5000.50), (2, 7500.50) RETURNING sum(balance);
driver: bad connection
root@:26257> panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x44fe3d7]

goroutine 94 [running]:
panic(0x51567c0, 0xc8200100b0)
	/usr/local/Cellar/go/1.6/libexec/src/runtime/panic.go:464 +0x3e6
github.com/cockroachdb/cockroach/sql/parser.(*FuncExpr).Eval(0xc820055f10, 0x1, 0xececbfb9c, 0x1aac21d2, 0x5e5a0c0, 0xececbfb9c, 0x1aabdffd, 0x5e5a0c0, 0x144f1cca9c26f508, 0x0, ...)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:1629 +0x367
github.com/cockroachdb/cockroach/sql.(*returningHelper).cookResultRow(0xc82081a548, 0xc820982660, 0x3, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/returning.go:97 +0x1fc
github.com/cockroachdb/cockroach/sql.(*insertNode).Next(0xc82081a540, 0xc82025a1e0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/insert.go:388 +0xe3e
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc820304000, 0x77a1198, 0xc820772c30, 0xc820399890, 0xc820772c01, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/executor.go:951 +0x40b
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtInOpenTxn(0xc820304000, 0x77a1198, 0xc820772c30, 0xc820399890, 0x101, 0xc820399828, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/executor.go:824 +0x545
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtsInCurrentTxn(0xc820304000, 0xc82086ebe0, 0x1, 0x1, 0xc820399890, 0xc820399828, 0x59c0101, 0x0, 0x0, 0x0, ...)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/executor.go:603 +0x5da
github.com/cockroachdb/cockroach/sql.runTxnAttempt(0xc820304000, 0xc820399890, 0x1, 0xc820399828, 0xc82070f220, 0xc82086ebe0, 0x1, 0x1, 0x0, 0x0, ...)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/executor.go:533 +0x18d
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest.func1(0xc82014adc0, 0xc82070f220, 0x0, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/executor.go:443 +0x34a
github.com/cockroachdb/cockroach/client.(*Txn).Exec(0xc82014adc0, 0x53d0101, 0x144f1cca9c26f508, 0x0, 0xc82099b4e0, 0x0, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/client/txn.go:532 +0x2eb
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc820304000, 0x77707d0, 0xc8203594c0, 0xc820399800, 0xc820819015, 0x5c, 0x0, 0x0, 0x0, 0x776d200)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/executor.go:448 +0x8e6
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc820304000, 0x77707d0, 0xc8203594c0, 0xc820399800, 0xc820819015, 0x5c, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/executor.go:337 +0xe5
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc820207200, 0x77707d0, 0xc8203594c0, 0xc820819015, 0x5c, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:646 +0xb2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc820207200, 0x77707d0, 0xc8203594c0, 0xc820207228, 0x0, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:323 +0xf2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc820207200, 0x0, 0x0, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:279 +0x1095
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc820328420, 0x6fad0a0, 0xc8202a9970, 0x0, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:226 +0x9e5
github.com/cockroachdb/cockroach/server.(*Server).Start.func6.1(0x6fad0a0, 0xc8202a9970)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/server/server.go:341 +0x3f
github.com/cockroachdb/cockroach/util.ServeHandler.func3.1(0xc820394200, 0x6fad0a0, 0xc8202a9970, 0xc820288010)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/util/net.go:125 +0x5a
created by github.com/cockroachdb/cockroach/util.ServeHandler.func3
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/util/net.go:127 +0x31f
```

### PR DESCRIPTION
**Pull request by dt**
Fixes #6705.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7052)

### CODE REVIEW

File: sql/returning.go [hunk 1]
```diff
@@ -50,6 +52,12 @@ func (p *planner) makeReturningHelper(
 		return rh, nil
 	}
 
+	for _, e := range r {
+		if p.aggregateInExpr(e.Expr) {
+			return rh, fmt.Errorf("aggregate functions are not allowed in RETURNING")
```
Comment (tamird) [inline sql/returning.go]: Is this error from postgres? I think we can do better by mentioning the specific illegal expression

**Comment by dt** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/returning.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052#-KJWz8YZYl68sBWDq5Iu-r1-57:-KJX-q_m899XS3yHYWli:666221560) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d272917aa1e74dc2f6b8e32d0be9cd0d15cbe32/sql/returning.go#L57)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Is this error from postgres? I think we can do better by mentioning the specific illegal expression
</details>
yeah, that's the pg error verbatim (though in the `psql`shell it echo's the whole query with a carrot under the offending clause, but this seems specific to the official shell).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052)*

**Comment by knz** [comment]
If we're hunting places where aggregates are not allowed, I'd suggest doing so for all in one swoop. I just tested, the following queries also crash the server:

- `SELECT 1 LIMIT COUNT(1)`   (LIMIT expressions: offset, count)
- `INSERT INTO foo VALUES (COUNT(1))`    (VALUES expressions)
- `create table aaa (y int,  constraint z check (count(1)>1)); insert into aaa values (1)` (CHECK expressions)
- `create table aac (x int default count(1), y int); insert into aac (y) values (1);`  (DEFAULT expressions)

Seems we've hit a nice schroedinbug here.

<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> #### sql: disallow aggregate funcs in RETURNING
> 
> Fixes #6705.
> 
> 
> 
</details>


---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052#-:-KJX4Jxwd2k_ssJFmvN9:1110161596)*

**Comment by dt** [comment]
I updated to get the easy ones (LIMIT, OFFSET, VALUES) but the exprs in the table desc will require a little more refactoring (I'm thinking I want to move the aggImpl's and the isAgg visitor to `parser`) so I'll file a bug for those to come back to them on another PR.

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> If we're hunting places where aggregates are not allowed, I'd suggest doing so for all in one swoop. I just tested, the following queries also crash the server:
> 
> - `SELECT 1 LIMIT COUNT(1)`   (LIMIT expressions: offset, count)
> - `INSERT INTO foo VALUES (COUNT(1))`    (VALUES expressions)
> - `create table aaa (y int,  constraint z check (count(1)>1)); insert into aaa values (1)` (CHECK expressions)
> - `create table aac (x int default count(1), y int); insert into aac (y) values (1);`  (DEFAULT expressions)
> 
> Seems we've hit a nice schroedinbug here.
</details>


---

Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052#-:-KJaw74vYb4grmGgD2Rt:559285885)*

**Comment by dt** [comment]
#7067

<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> I updated to get the easy ones (LIMIT, OFFSET, VALUES) but the exprs in the table desc will require a little more refactoring (I'm thinking I want to move the aggImpl's and the isAgg visitor to `parser`) so I'll file a bug for those to come back to them on another PR.
</details>


---

Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052#-:-KJawg94xi9IEQYkhtVf:707971826)*

**Comment by knz** [comment]
:lgtm:

<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> #7067
</details>


---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052#-:-KJaz0fkt3XDcE2Du4wP:1417088721)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/returning.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052#-KJWz8YZYl68sBWDq5Iu-r1-57:-KJazXgrCjsrjsAGP4R8:-1775796546) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d272917aa1e74dc2f6b8e32d0be9cd0d15cbe32/sql/returning.go#L57)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> yeah, that's the pg error verbatim (though in the `psql`shell it echo's the whole query with a carrot under the offending clause, but this seems specific to the official shell).
</details>
Wait, so why did we decide not to emit a better error?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/returning.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052#-KJWz8YZYl68sBWDq5Iu-r1-57:-KJazrjDimPDNBq6Pb_6:1473629428) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d272917aa1e74dc2f6b8e32d0be9cd0d15cbe32/sql/returning.go#L57)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Wait, so why did we decide not to emit a better error?
</details>
I think that's because we need to fix #2037 first.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/returning.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052#-KJWz8YZYl68sBWDq5Iu-r1-57:-KJb-HMqfhq8toaiWZ6W:-1924273787) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d272917aa1e74dc2f6b8e32d0be9cd0d15cbe32/sql/returning.go#L57)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I think that's because we need to fix #2037 first.
</details>
That's one solution, but you could also simply include `e` in the message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7052)*

**Pull request merged**